Firestore/core/src/firebase/firestore/util/filesystem_win.cc

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+
+#include <windows.h>
+
+#include <cerrno>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+bool Exists(const Path& path, bool* is_directory) {
+  DWORD attrs = ::GetFileAttributesW(path.c_str());
+  if (attrs == INVALID_FILE_ATTRIBUTES) {
+    return false;
+  }
+
+  if (is_directory) {
+    *is_directory = attrs & FILE_ATTRIBUTE_DIRECTORY;
+  }
+  return true;
+}
+
+Path TempDir() {
+  // Returns a null-terminated string with a trailing backslash.
+  wchar_t buffer[MAX_PATH + 1];
+  DWORD count = ::GetTempPathW(MAX_PATH, buffer);
+  HARD_ASSERT(count > 0, "Failed to determine temporary directory %s",
+              ::GetLastError());
+
+  return Path::FromUtf16(buffer, count);
+}
+
+namespace detail {
+
+Status CreateDir(const Path& path) {
+  if (::CreateDirectoryW(path.c_str(), nullptr)) {
```

Comment:
The docs claim that path.c_str() must be less than 248 chars (or ~32k if it starts with \?). But the docs don't elaborate on what happens if this pre-condition isn't met. Presumably, the function fails with some reasonable error code. So this is probably ok.

If you want to be paranoid, you could hard_assert the length first (or return a failed status if you want to be more friendly.) Doing so properly would be tricky though, since the length differs based on the prefix of the string OR on a setting in windows 10+. Pragmatically, you're probably better off doing what you're doing and just catching the error code and presuming that Microsoft does something reasonable in this situation.